Mathlib/Order/Minimal.lean

```diff
@@ -302,28 +302,98 @@ section Image
 
 variable {f : α → β} {r : α → α → Prop} {s : β → β → Prop}
 
-theorem minimals_image_of_rel_iff_rel (hf : ∀ ⦃a a'⦄, a ∈ x → a' ∈ x → (r a a' ↔ s (f a) (f a'))) :
-    minimals s (f '' x) = f '' (minimals r x) := by
-  ext a
-  simp only [minimals, mem_image, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂]
-  constructor
-  · rintro ⟨⟨a, ha, rfl⟩ , h⟩
-    exact ⟨a, ⟨ha, fun y hy hya ↦ (hf ha hy).mpr (h _ hy ((hf hy ha).mp hya))⟩, rfl⟩
-  rintro ⟨a,⟨⟨ha,h⟩,rfl⟩⟩
-  exact ⟨⟨_, ha, rfl⟩, fun y hy hya ↦ (hf ha hy).mp (h hy ((hf hy ha).mpr hya))⟩
-
-theorem maximals_image_of_rel_iff_rel_on
-    (hf : ∀ ⦃a a'⦄, a ∈ x → a' ∈ x → (r a a' ↔ s (f a) (f a'))) :
-    maximals s (f '' x) = f '' (maximals r x) :=
-  minimals_image_of_rel_iff_rel (fun _ _ a_1 a_2 ↦ hf a_2 a_1)
+section
+variable {x : Set α} (hf : ∀ ⦃a a'⦄, a ∈ x → a' ∈ x → (r a a' ↔ s (f a) (f a')))
+
+theorem map_mem_minimals {a : α} (ha : a ∈ minimals r x) : f a ∈ minimals s (f '' x) :=
+  ⟨⟨a, ha.1, rfl⟩, by rintro _ ⟨a', h', rfl⟩; rw [← hf ha.1 h', ← hf h' ha.1]; exact ha.2 h'⟩
+
+theorem map_mem_minimals_iff {a : α} (ha : a ∈ x) : f a ∈ minimals s (f '' x) ↔ a ∈ minimals r x :=
+  ⟨fun ⟨_, hmin⟩ ↦ ⟨ha, fun a' h' ↦ by
+    simpa only [hf h' ha, hf ha h'] using hmin ⟨a', h', rfl⟩⟩, map_mem_minimals hf⟩
+
+theorem minimals_image_of_rel_iff_rel : minimals s (f '' x) = f '' (minimals r x) := by
+  ext b; refine ⟨fun h ↦ ?_, ?_⟩
+  · obtain ⟨a, ha, rfl⟩ := h.1; exact ⟨a, (map_mem_minimals_iff hf ha).mp h, rfl⟩
+  · rintro ⟨a, ha, rfl⟩; exact map_mem_minimals hf ha
+
+theorem map_mem_maximals {a : α} (ha : a ∈ maximals r x) : f a ∈ maximals s (f '' x) :=
+  map_mem_minimals (fun _ _ h₁ h₂ ↦ by exact hf h₂ h₁) ha
+
+theorem map_mem_maximals_iff {a : α} (ha : a ∈ x) : f a ∈ maximals s (f '' x) ↔ a ∈ maximals r x :=
+  map_mem_minimals_iff (fun _ _ h₁ h₂ ↦ by exact hf h₂ h₁) ha
+
+theorem maximals_image_of_rel_iff_rel : maximals s (f '' x) = f '' (maximals r x) :=
+  minimals_image_of_rel_iff_rel fun _ _ h₁ h₂ ↦ hf h₂ h₁
```

Comment:
Do you mind keeping the analogous `minimals` and `maximals` results just after each other?

Reply:
Done!